### PR TITLE
Remove Quadrature(unsigned int) from the public interface.

### LINKED ISSUE #16818: Remove constructor Quadrature(unsigned int).
My exchange student wanted `QGauss<dim>(3)`, but accidentally did `Quadrature<dim>(3)`. This creates a quadrature formula with three points, all at the origin, and weight zero. Perhaps not surprisingly, in hindsight, all his integrals were zero. It was not what he wanted.

I think we should remove this constructor, or at least make it `protected`. It doesn't create a useful object, and leads to hard-to-spot mistakes.

### PR DESCRIPTION
**Pull request by bangerth**
Fixes #16818.

This is strictly speaking an incompatible change. I don't think there should be many places where someone creates a correctly sized but otherwise invalid quadrature object, so the impact is likely small to zero.

I did think about alternatives. The obvious one is the introduction of a default constructor `Quadrature()` and removing the default value from the constructor in question here, followed by marking the latter as deprecated. That could work to retain backward compatibility. But because derived classes uniformly seem to be using this constructor, once we remove deprecated functions, we have to remember that this one cannot be removed but must be made `protected` instead. That just seemed unnecessary use of brain power, so I decided that the impact here is small enough to warrant outright removal.

### CODE REVIEW

**Review by kronbichler — APPROVED** [review]

**Comment by kronbichler** [comment]
We need to address the error
```
/home/runner/work/dealii/dealii/include/deal.II/matrix_free/fe_remote_evaluation.h:1192:19: error: calling a protected constructor of class 'dealii::Quadrature<2>' [clang-diagnostic-error]
```
This was recently introduced by @jh66637 in #16394 - any comments on how to address this? I think it would make sense to not store quadrature formulas but merely points (depending on how the consumer site looks like), or fill the quadrature objects with one of the other constructors.

**Comment by bangerth** [comment]
The code you point out looks like this:
```
                // Insert a quadrature rule of correct size into the global
                // quadrature vector. First check that each face is only
                // considered once.
                Assert(global_quadrature_vector[bface].size() == 0,
                       ExcMessage(
                         "Quadrature for given face already provided."));

                global_quadrature_vector[bface] =
                  Quadrature<dim>(phi.n_q_points);
```
I think it's worth pointing out that what one ends up with here is a quadrature object with `phi.n_q_point` quadrature points that are all at the origin, and zero weights. This is a prescription for bugs (which is how I found out about this constructor to begin with). In other places we would initialize points and weights with `signaling_nan`s, but here we don't. 

In other words, I think that the design used here is broken. The object as initialized is unusable. We shouldn't create it. I don't know where that is, but assume that points and weights are set somewhere else, and we should leave resizing to the correct size to the place where we do the initialization of points and weights.

I'm afraid I have no overview of this code. Could someone help me with this location?

**Comment by kronbichler** [comment]
I can definitely help if @jh66637 is busy, let us wait for a day or two (as my schedule is overly crowded today and tomorrow).

File: doc/news/changes/incompatibilities/20240513Bangerth-2 [hunk 1]
```diff
@@ -0,0 +1,14 @@
+Changed: The Quadrature class used to have a constructor that took an
+integer argument. This was error prone because it was easy to accidentally
```
Comment (masterleinad) [inline doc/news/changes/incompatibilities/20240513Bangerth-2:2]:
```suggestion
Changed: The Quadrature class used to have a constructor that took an
integer argument. This was error-prone because it was easy to accidentally
```

**Review by masterleinad — APPROVED** [review]

**Comment by jh66637** [comment]
@bangerth @kronbichler The reason I used this is the following: We only need the size of the quadrature objects. At some point I pointed out, that we have to decide if we keep the interface similar to `FERemoteEvaluationCommunicator::reinit_faces()` which requires a vector of quadratures or simply hand over a vector of quadrature sizes. I didn't find the respective comment, but at that time there was no opinion on how to proceed, and I kept it the way the function was written initially. 

Long story short, we can simply replace
```
FERemoteEvaluationCommunicator::reinit_faces(..., const std::vector<std::vector<Quadrature<dim - 1>>> &quadrature_vector)
```
by
```
FERemoteEvaluationCommunicator::reinit_faces(..., const std::vector<std::vector<unsigned int>> &quadrature_sizes)
```
and similar for all overloads. This functionality is quite new, was in no release and is probably not used by a lot of people, so this can be easily changed in my opinion. I can do that in a separate PR if you want to, or you can do it with this PR. What do you prefer?

**Comment by bangerth** [comment]
Since you know the code, I'd love if you could make that change in a separate PR. Once that's merged, I can rebase the current one. Thank you in advance!

**Comment by jh66637** [comment]
> Since you know the code, I'd love if you could make that change in a separate PR. 
I will try to find the time tomorrow :)

**Comment by bangerth** [comment]
Now ready given that the one place where this failed has been addressed in #17032.

**Comment by masterleinad** [comment]
There are still failing tests: 
> The following tests FAILED:
	 61 - simplex/data_out_write_gnuplot_01.debug (Failed)
	 62 - simplex/data_out_write_gnuplot_02.debug (Failed)
	 63 - simplex/data_out_write_gnuplot_03.debug (Failed)
	 68 - simplex/data_out_write_vtk_01.debug (Failed)
	 69 - simplex/data_out_write_vtk_02.debug (Failed)
	155 - simplex/step-23.debug (Failed)
	156 - simplex/step-31.debug (Failed)
	163 - test_dependency/simplex.step-67.debug.executable (Failed)
	164 - simplex/step-67.mpirun=1.debug (Not Run)
	165 - simplex/step-67.mpirun=2.debug (Not Run)

**Review by kronbichler — APPROVED** [review]

**Comment by bangerth** [comment]
I made it to work now.

File: tests/base/quadrature_is_tensor_product.cc [hunk 2]
```diff
@@ -26,7 +26,8 @@ template <int dim>
 void
 print_is_tensor_product()
 {
-  Quadrature<dim> q_0(1);
+  Quadrature<dim> q_0(std::vector<Point<dim>>{Point<dim>()},
+                      std::vector<double>{0.});
```
Comment (marcfehling) [inline tests/base/quadrature_is_tensor_product.cc:30]:
```suggestion
  Quadrature<dim> q_0;
```

File: tests/hp/fe_nothing_dominates_01.cc [hunk 3]
```diff
@@ -46,7 +46,10 @@ test(const bool fe_nothing_dominates)
     hp::FECollection<dim> fe_collection(FE_Nothing<dim>(/*n_components=*/1,
                                                         fe_nothing_dominates),
                                         FE_Q<dim>(1));
-    hp::QCollection<dim>  q_collection(Quadrature<dim>(1), QGauss<dim>(2));
+    hp::QCollection<dim>  q_collection(
+      Quadrature<dim>(std::vector<Point<dim>>{Point<dim>()},
+                      std::vector<double>{0.}),
```
Comment (marcfehling) [inline tests/hp/fe_nothing_dominates_01.cc:51]:
```suggestion
      Quadrature<dim>(),
```

File: tests/hp/fe_nothing_dominates_01.cc [hunk 4]
```diff
@@ -55,7 +58,10 @@ test(const bool fe_nothing_dominates)
     hp::FECollection<dim> fe_collection(FE_Q<dim>(1),
                                         FE_Nothing<dim>(/*n_components=*/1,
                                                         fe_nothing_dominates));
-    hp::QCollection<dim>  q_collection(QGauss<dim>(2), Quadrature<dim>(1));
+    hp::QCollection<dim>  q_collection(
+      QGauss<dim>(2),
+      Quadrature<dim>(std::vector<Point<dim>>{Point<dim>()},
+                      std::vector<double>{0.}));
```
Comment (marcfehling) [inline tests/hp/fe_nothing_dominates_01.cc:64]:
```suggestion
      Quadrature<dim>());
```

File: tests/hp/fe_nothing_dominates_02.cc [hunk 5]
```diff
@@ -48,15 +48,21 @@ test()
     hp::FECollection<dim> fe_collection(
       FESystem<dim>(FE_Q<dim>(1), FE_Q<dim>(1)),
       FESystem<dim>(FE_Nothing<dim>(1, true), FE_Nothing<dim>(1, false)));
-    hp::QCollection<dim> q_collection(QGauss<dim>(2), Quadrature<dim>(1));
+    hp::QCollection<dim> q_collection(
+      QGauss<dim>(2),
+      Quadrature<dim>(std::vector<Point<dim>>{Point<dim>()},
+                      std::vector<double>{0.}));
```
Comment (marcfehling) [inline tests/hp/fe_nothing_dominates_02.cc:54]:
```suggestion
      Quadrature<dim>());
```

File: tests/hp/fe_nothing_dominates_02.cc [hunk 6]
```diff
@@ -48,15 +48,21 @@ test()
...
   }
   {
     deallog << "(FE_Nothing(true)xFE_Nothing(false), FE_QxFE_Q)" << std::endl;
     hp::FECollection<dim> fe_collection(
       FESystem<dim>(FE_Nothing<dim>(1, true), FE_Nothing<dim>(1, false)),
       FESystem<dim>(FE_Q<dim>(1), FE_Q<dim>(1)));
-    hp::QCollection<dim> q_collection(Quadrature<dim>(1), QGauss<dim>(2));
+    hp::QCollection<dim> q_collection(
+      Quadrature<dim>(std::vector<Point<dim>>{Point<dim>()},
+                      std::vector<double>{0.}),
```
Comment (marcfehling) [inline tests/hp/fe_nothing_dominates_02.cc:64]:
```suggestion
      Quadrature<dim>(),
```

**Review by marcfehling — APPROVED** [review]
One less pitfall. Very nice!

You can also use the default constructor in the tests.

File: tests/base/quadrature_is_tensor_product.cc (hunk 2 above)
Comment (bangerth) [reply tests/base/quadrature_is_tensor_product.cc:30]: No, not the same. I'm creating a quadrature formula with one point located at the origin and weight zero. You're suggesting to create a quadrature formula with no points.

(I have no idea why the author of the original test chose such a quadrature object, but that's what the test does and I did not feel like investigating whether that's something useful or an accident.)

Comment (marcfehling) [reply tests/base/quadrature_is_tensor_product.cc:30]: Ah right. I didn't pay enough attention to the initializer lists. Mea culpa.

**Review by marcfehling — APPROVED** [review]
